pruning/manager.go

```diff
@@ -0,0 +1,282 @@
+// by calling SetOptions.
+func NewManager(db dbm.DB, logger log.Logger) *Manager {
+	return &Manager{
+		db:                   db,
+		logger:               logger,
+		opts:                 types.NewPruningOptions(types.PruningNothing),
+		pruneHeights:         []int64{},
+		pruneSnapshotHeights: list.New(),
+	}
+}
+
+// SetOptions sets the pruning strategy on the manager.
+func (m *Manager) SetOptions(opts types.PruningOptions) {
+	m.opts = opts
+}
+
+// GetOptions fetches the pruning strategy from the manager.
+func (m *Manager) GetOptions() types.PruningOptions {
+	return m.opts
+}
+
+// GetFlushAndResetPruningHeights returns all heights to be pruned during the next call to Prune().
+// It also flushes and resets the pruning heights.
+func (m *Manager) GetFlushAndResetPruningHeights() ([]int64, error) {
+	if m.opts.GetPruningStrategy() == types.PruningNothing {
+		return []int64{}, nil
+	}
+	m.pruneHeightsMx.Lock()
+	defer m.pruneHeightsMx.Unlock()
+
+	// flush the updates to disk so that it is not lost if crash happens.
+	if err := m.db.SetSync(pruneHeightsKey, int64SliceToBytes(m.pruneHeights)); err != nil {
+		return nil, err
+	}
+
+	// Return a copy to prevent data races.
+	pruningHeights := make([]int64, len(m.pruneHeights))
+	copy(pruningHeights, m.pruneHeights)
+	m.pruneHeights = m.pruneHeights[:0]
+
+	return pruningHeights, nil
+}
+
+// HandleHeight determines if pruneHeight height needs to be kept for pruning at the right interval prescribed by
+// the pruning strategy. Returns previousHeight, if it was kept to be pruned at the next call to Prune(), 0 otherwise.
+// previousHeight must be greater than 0 for the handling to take effect since valid heights start at 1 and 0 represents
+// the latest height. The latest height cannot be pruned. As a result, if previousHeight is less than or equal to 0, 0 is returned.
+func (m *Manager) HandleHeight(previousHeight int64) int64 {
+	if m.opts.GetPruningStrategy() == types.PruningNothing || previousHeight <= 0 {
+		return 0
+	}
+
+	defer func() {
+		m.pruneHeightsMx.Lock()
+		defer m.pruneHeightsMx.Unlock()
+
+		m.pruneSnapshotHeightsMx.Lock()
+		defer m.pruneSnapshotHeightsMx.Unlock()
+
+		// move persisted snapshot heights to pruneHeights which
+		// represent the heights to be pruned at the next pruning interval.
+		var next *list.Element
+		for e := m.pruneSnapshotHeights.Front(); e != nil; e = next {
+			snHeight := e.Value.(int64)
+			if snHeight < previousHeight-int64(m.opts.KeepRecent) {
+				m.pruneHeights = append(m.pruneHeights, snHeight)
+
+				// We must get next before removing to be able to continue iterating.
+				next = e.Next()
+				m.pruneSnapshotHeights.Remove(e)
+			} else {
+				next = e.Next()
+			}
+		}
+
+		// flush the updates to disk so that they are not lost if crash happens.
+		if err := m.db.SetSync(pruneHeightsKey, int64SliceToBytes(m.pruneHeights)); err != nil {
+			panic(err)
+		}
+	}()
+
+	if int64(m.opts.KeepRecent) < previousHeight {
```

Comment:
why is a height being compared to an interval?

Reply:
`keep-recent` represents the number of heights to keep relative to the previous height

Let's assume that `keep-recent` is 50, `previousHeight=50`, then we do nothing.

During the next commit, `previousHeight` becomes 51. So we have `pruneHeight = 51 - 50 = 1`. Therefore, we add 1 to the list of heights to be pruned